web3.js/src/connection.ts

```diff
@@ -4263,75 +4207,273 @@ export class Connection {
       return;
     }
 
-    for (let clientSubscriptionId of accountClientSubscriptionIds) {
-      const sub = this._accountChangeSubscriptions[clientSubscriptionId];
-      this._subscribe(
-        sub,
-        'accountSubscribe',
-        this._buildArgs([sub.publicKey], sub.commitment, 'base64'),
-      );
-    }
-
-    for (let clientSubscriptionId of programClientSubscriptionIds) {
-      const sub = this._programAccountChangeSubscriptions[clientSubscriptionId];
-      this._subscribe(
-        sub,
-        'programSubscribe',
-        this._buildArgs([sub.programId], sub.commitment, 'base64', {
-          filters: sub.filters,
-        }),
-      );
-    }
-
-    for (let clientSubscriptionId of slotClientSubscriptionIds) {
-      const sub = this._slotSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'slotSubscribe', []);
-    }
-
-    for (let clientSubscriptionId of slotUpdateClientSubscriptionIds) {
-      const sub = this._slotUpdateSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'slotsUpdatesSubscribe', []);
-    }
-
-    for (let clientSubscriptionId of signatureClientSubscriptionIds) {
-      const sub = this._signatureSubscriptions[clientSubscriptionId];
-      const args: any[] = [sub.signature];
-      if (sub.options) args.push(sub.options);
-      this._subscribe(sub, 'signatureSubscribe', args);
-    }
+    await Promise.all(
+      Object.entries(
+        this._subscriptionsByHash as Record<
+          SubscriptionConfigHash,
+          Subscription
+        >,
+      ).map(async ([hash, subscription]) => {
+        switch (subscription.state) {
+          case 'pending':
+          case 'unsubscribed':
+            if (subscription.callbacks.size === 0) {
+              /**
+               * You can end up here when:
+               *
+               * - a subscription has recently unsubscribed
+               *   without having new callbacks added to it
+               *   while the unsubscribe was in flight, or
+               * - when a pending subscription has its
+               *   listeners removed before a request was
+               *   sent to the server.
+               *
+               * Being that nobody is interested in this
+               * subscription any longer, delete it.
+               */
+              delete this._subscriptionsByHash[hash];
+              if (subscription.state === 'unsubscribed') {
+                delete this._subscriptionCallbacksByServerSubscriptionId[
+                  subscription.serverSubscriptionId
+                ];
+              }
+              await this._updateSubscriptions();
```

Comment:
It worries me that one invocation of `_updateSubscriptions` could trigger many other recursive updates. The top level invocation and the lower level invocations would all be running at the same time and could cause all sorts of race conditions.

Reply:
OMG. I wrote a gigantic explainer here. Now, it's gone. Damnit GitHub.

Reply:
> The top level invocation and the lower level invocations would all be running at the same time…

But not _actually_ at the same time, right? The way this state machine is written, every state transition is executed synchronously and eagerly. Every time a subscription enters the state machine it either transitions or skips. So long as every pass through the machine causes a transition, there's no risk of infinite recursion, and because every subscription operates independently of the others, there are no races.

When you encounter a recursion, those also get processed synchronously and eagerly. Imagine that you had 4 subscriptions in the `pending` state, and the first one happened to have zero callbacks. The stack would look something like this.

```
1A (no callbacks; gets deleted then recurses)
 L 2B (transitions to subscribing)
 L 3B (transitions to subscribing)
 L 4B (transitions to subscribing) 
2A (subscribing, so skips)
3A (subscribing, so skips)
4A (subscribing, so skips)
```

One nice thing about going through the entire update function recursively is that, as the state whittles itself down to a stable state, you get all of the socket timeout connection/disconnection logic for free.

Reply:
Ok, this flow makes sense and is well thought out, thanks for the explanation!